### PR TITLE
NAS-106859 / 12.1 / Nas 106859

### PR DESCRIPTION
**Pull request by dkmullen**
On SCALE:
- Enables VNC Settings for either bootloader options but...
- Removes Delay VM Boot Until VNC Connects
- Doesn't offer Grub as an option

On Core, Enterprise
- hides VNC options for Grub and Legacy

On all:
- Exposes verbose bootloader descriptions to the user

### CODE REVIEW

File: src/app/pages/vm/vm-wizard/vm-wizard.component.ts [hunk 1]
```diff
@@ -874,9 +880,20 @@ async customSubmit(value) {
...
-    
-    if(value.enable_vnc &&value.bootloader !== "UEFI_CSM"){
-      vm_payload["devices"].push({
+
+    if (value.enable_vnc) {
+      if (this.productType === 'SCALE') {
+        vm_payload["devices"].push({
+          "dtype": "VNC", "attributes": {
+            "vnc_port": String(this.getRndInteger(5900,65535)),
+            "vnc_resolution": "1024x768",
```
Comment (sonicaj) [inline src/app/pages/vm/vm-wizard/vm-wizard.component.ts:889]: Setting resolution is not supported in SCALE. Can we please remove this ?

File: src/app/pages/vm/vm-wizard/vm-wizard.component.ts [hunk 2]
```diff
@@ -886,7 +886,6 @@ async customSubmit(value) {
         vm_payload["devices"].push({
           "dtype": "VNC", "attributes": {
             "vnc_port": String(this.getRndInteger(5900,65535)),
-            "vnc_resolution": "1024x768",
```
Comment (erinix) [inline src/app/pages/vm/vm-wizard/vm-wizard.component.ts]: do we want to keep it in conditionally for 12.1/freebsd? (I notice there is a backport of this too, does it make sense to port any of these changes back to 12?)

Comment (dkmullen) [reply src/app/pages/vm/vm-wizard/vm-wizard.component.ts]: I kept it in for 12.1 and 12 ( few lines below the above snippet). The change to middleware affects CORE and ENTERPRISE too - https://github.com/freenas/freenas/pull/5229

Comment (erinix) [reply src/app/pages/vm/vm-wizard/vm-wizard.component.ts]: I still don't see where you are keeping the resolution, what line is that on?

File: src/app/pages/vm/vm-wizard/vm-wizard.component.ts [hunk 3]
```diff
@@ -874,9 +880,19 @@ async customSubmit(value) {
...
     }
-    
-    if(value.enable_vnc &&value.bootloader !== "UEFI_CSM"){
-      vm_payload["devices"].push({
+
+    if (value.enable_vnc) {
+      if (this.productType === 'SCALE') {
+        vm_payload["devices"].push({
+          "dtype": "VNC", "attributes": {
+            "vnc_port": String(this.getRndInteger(5900,65535)),
```
Comment (sonicaj) [inline src/app/pages/vm/vm-wizard/vm-wizard.component.ts:888]: I just noticed, is this a custom function retrieve a random integer between this range ?
If yes, we should not be doing this and instead we should make a call to middleware which will return the next available vnc port `vm.vnc_port_wizard`.

File: src/app/pages/vm/vm-wizard/vm-wizard.component.ts (hunk 2 above)
Comment (dkmullen) [reply src/app/pages/vm/vm-wizard/vm-wizard.component.ts]: It's on line 903. The conditional on 898 runs if the system isn't SCALE, adds in the non-SCALE properties of 'wait' and 'resolution'

Comment (erinix) [reply src/app/pages/vm/vm-wizard/vm-wizard.component.ts]: Oh I see now, sorry about that

**Pull request merged**